client/pd_service_discovery.go

```diff
@@ -1145,20 +1116,47 @@ func (c *pdServiceDiscovery) switchTSOAllocatorLeaders(allocatorMap map[string]*
 	return nil
 }
 
-// GetOrCreateGRPCConn returns the corresponding grpc client connection of the given addr
-func (c *pdServiceDiscovery) GetOrCreateGRPCConn(addr string) (*grpc.ClientConn, error) {
-	return grpcutil.GetOrCreateGRPCConn(c.ctx, &c.clientConns, addr, c.tlsCfg, c.option.gRPCDialOptions...)
+// GetOrCreateGRPCConn returns the corresponding grpc client connection of the given URL.
+func (c *pdServiceDiscovery) GetOrCreateGRPCConn(url string) (*grpc.ClientConn, error) {
+	return grpcutil.GetOrCreateGRPCConn(c.ctx, &c.clientConns, url, c.tlsCfg, c.option.gRPCDialOptions...)
 }
 
-func addrsToUrls(addrs []string) []string {
+func addrsToURLs(addrs []string, tlsCfg *tls.Config) []string {
 	// Add default schema "http://" to addrs.
 	urls := make([]string, 0, len(addrs))
 	for _, addr := range addrs {
-		if strings.Contains(addr, "://") {
-			urls = append(urls, addr)
-		} else {
-			urls = append(urls, "http://"+addr)
-		}
+		urls = append(urls, addrToURL(addr, tlsCfg))
 	}
 	return urls
 }
+
+func addrToURL(addr string, tlsCfg *tls.Config) string {
+	if tlsCfg == nil {
+		if strings.HasPrefix(addr, httpsScheme) {
+			addr = fmt.Sprintf("%s%s", httpScheme, strings.TrimPrefix(addr, httpsScheme))
+		} else if !strings.HasPrefix(addr, httpScheme) {
+			addr = fmt.Sprintf("%s%s", httpScheme, addr)
+		}
+	} else {
+		if strings.HasPrefix(addr, httpScheme) {
+			addr = fmt.Sprintf("%s%s", httpsScheme, strings.TrimPrefix(addr, httpScheme))
+		} else if !strings.HasPrefix(addr, httpsScheme) {
+			addr = fmt.Sprintf("%s%s", httpsScheme, addr)
+		}
+	}
+	return addr
+}
+
+// pickMatchedURL picks the matched URL based on the TLS config.
+func pickMatchedURL(urls []string, tlsCfg *tls.Config) string {
+	for _, url := range urls {
+		if tlsCfg != nil && strings.HasPrefix(url, httpsScheme) {
+			return url
+		}
+		if tlsCfg == nil && strings.HasPrefix(url, httpScheme) {
+			return url
+		}
+	}
+	log.Warn("[pd] no matched url found", zap.Strings("urls", urls), zap.Bool("tls-enabled", tlsCfg != nil))
+	return urls[0]
```

Comment:
Do we need to change the URL according to the tlscfg instead of return [0]?

Reply:
Good catch.